params/config.go

```diff
@@ -347,72 +333,95 @@ type XDPoSConfig struct {
 }
 
 type V2 struct {
-	FirstSwitchBlock *big.Int             `json:"switchBlock"`
-	CurrentConfig    *V2Config            `json:"config"`
-	AllConfigs       map[uint64]*V2Config `json:"allConfigs"`
-	configIndex      []uint64             //list of switch block of configs
-	SkipV2Validation bool                 //Skip Block Validation for testing purpose, V2 consensus only
+	lock sync.RWMutex // Protects the signer fields
+
+	SwitchBlock   *big.Int             `json:"switchBlock"`
+	CurrentConfig *V2Config            `json:"config"`
+	AllConfigs    map[uint64]*V2Config `json:"allConfigs"`
+	configIndex   []uint64             //list of switch block of configs
+
+	SkipV2Validation bool //Skip Block Validation for testing purpose, V2 consensus only
 }
 
 type V2Config struct {
-	WaitPeriod           int      `json:"waitPeriod"`           // Miner wait period to check mine event
-	MinePeriod           int      `json:"minePeriod"`           // Miner mine period to mine a block
-	SwitchBlock          *big.Int `json:"switchBlock"`          // v1 to v2 switch block number
-	TimeoutSyncThreshold int      `json:"timeoutSyncThreshold"` // send syncInfo after number of timeout
-	TimeoutPeriod        int      `json:"timeoutPeriod"`        // Duration in ms
-	CertThreshold        int      `json:"certificateThreshold"` // Necessary number of messages from master nodes to form a certificate
+	SwitchRound          uint64 `json:"switchRound"`          // v1 to v2 switch block number
+	WaitPeriod           int    `json:"waitPeriod"`           // Miner wait period to check mine event
+	MinePeriod           int    `json:"minePeriod"`           // Miner mine period to mine a block
+	TimeoutSyncThreshold int    `json:"timeoutSyncThreshold"` // send syncInfo after number of timeout
+	TimeoutPeriod        int    `json:"timeoutPeriod"`        // Duration in ms
+	CertThreshold        int    `json:"certificateThreshold"` // Necessary number of messages from master nodes to form a certificate
 }
 
-// String implements the stringer interface, returning the consensus engine details.
-func (c *XDPoSConfig) BuildConfigIndex() {
-	var list []uint64
+func (c *XDPoSConfig) String() string {
+	return "XDPoS"
+}
 
-	for i := range c.V2.AllConfigs {
-		list = append(list, i)
+func (c *XDPoSConfig) BlockConsensusVersion(num *big.Int, extraByte []byte, extraCheck bool) string {
+	if extraCheck && (len(extraByte) == 0 || extraByte[0] != 2) {
+		return ConsensusEngineVersion1
```

Comment:
The consensus Version check shall purely based on the block number or round. Why does it need to know the `extraByte`?

Reply:
It's related to this PR. https://github.com/XinFinOrg/XDPoSChain/pull/186

Reply:
If we don't put this, synchronization on the special case mentioned in the PR will fail.